pkg/registry/common/refresh/nse_registry_client.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package refresh
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+)
+
+type refreshNSEClient struct {
+	client    registry.NetworkServiceEndpointRegistryClient
+	nsesMutex sync.Mutex
+	nses      map[string]func()
+}
+
+func (c *refreshNSEClient) startRefresh(ctx context.Context, nse *registry.NetworkServiceEndpoint) {
+	t := time.Unix(nse.ExpirationTime.Seconds, int64(nse.ExpirationTime.Nanos))
+	delta := time.Until(t)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(2 * time.Until(t) / 3):
+				t1 := time.Now().Add(delta)
+				nse.ExpirationTime.Seconds = t1.Unix()
+				nse.ExpirationTime.Nanos = int32(t1.Nanosecond())
+				var err error
+				nse, err = c.client.Register(ctx, nse)
+				if err != nil {
+					continue
```

Comment:
We need to have retry and some delay between them.

Reply:
Fixed